src/core/server/elasticsearch/elasticsearch_service.ts

```diff
@@ -0,0 +1,140 @@
+ */
+
+import { ConnectableObservable, Observable, Subscription } from 'rxjs';
+import { filter, first, publishReplay, switchMap } from 'rxjs/operators';
+import { CoreContext, CoreService } from '../../types';
+import { Logger } from '../logging';
+import { ClusterClient } from './cluster_client';
+import { ElasticsearchClientConfig } from './elasticsearch_client_config';
+import { ElasticsearchConfig } from './elasticsearch_config';
+
+interface CoreClusterClients {
+  config: ElasticsearchConfig;
+  adminClient: ClusterClient;
+  dataClient: ClusterClient;
+}
+
+/** @internal */
+export interface ElasticsearchServiceStartContract {
+  // Required for the BWC only.
+  readonly bwc: {
+    readonly config: ElasticsearchConfig;
+  };
+
+  readonly apiVersion: ElasticsearchConfig['apiVersion'];
+  readonly requestTimeout: ElasticsearchConfig['requestTimeout'];
+  readonly shardTimeout: ElasticsearchConfig['shardTimeout'];
+
+  readonly createClient: (
+    type: string,
+    config?: Partial<ElasticsearchClientConfig>
+  ) => ClusterClient;
+  readonly adminClient: ClusterClient;
+  readonly dataClient: ClusterClient;
+}
+
+/** @internal */
+export class ElasticsearchService implements CoreService<ElasticsearchServiceStartContract> {
+  private readonly log: Logger;
+  private subscription?: Subscription;
+
+  constructor(private readonly coreContext: CoreContext) {
+    this.log = coreContext.logger.get('elasticsearch-service');
+  }
+
+  public async start(): Promise<ElasticsearchServiceStartContract> {
+    this.log.debug('Starting elasticsearch service');
+
+    const clients$ = this.coreContext.configService
+      .atPath('elasticsearch', ElasticsearchConfig)
+      .pipe(
+        filter(() => {
+          if (this.subscription !== undefined) {
+            this.log.error('Clients cannot be changed after they are created');
+            return false;
+          }
+
+          return true;
+        }),
+        switchMap(
+          config =>
+            new Observable<CoreClusterClients>(subscriber => {
+              this.log.debug(`Creating elasticsearch clients`);
+
+              const coreClients = {
+                config,
+                adminClient: this.createClusterClient('admin', config),
+                dataClient: this.createClusterClient('data', config),
+              };
+
+              subscriber.next(coreClients);
+
+              return () => {
+                this.log.debug(`Closing elasticsearch clients`);
+
+                coreClients.adminClient.close();
+                coreClients.dataClient.close();
+              };
+            })
+        ),
+        publishReplay(1)
+      ) as ConnectableObservable<CoreClusterClients>;
+
+    this.subscription = clients$.connect();
+
+    const { config: defaultConfig, adminClient, dataClient } = await clients$
+      .pipe(first())
+      .toPromise();
+
+    return {
+      bwc: { config: defaultConfig },
+
+      apiVersion: defaultConfig.apiVersion,
```

Comment:
__note:__ these for ES plugin that expose these fields to the client side, also timelion uses `shardTimeout` for some reason.